[DOCS] [GSoC] add docs about how to build OV on RISC-V devices (#26056)
openvinotoolkit/openvino: docs/dev/build_riscv64.md

Reviewer: could you please mention this document there as well https://github.com/openvinotoolkit/openvino/blob/master/docs/dev/build.md ?

Author: Sure!

# Cross compile OpenVINO™ Runtime for RISCV64 systems

> This guide shows how to build OpenVINO Runtime for 64-bit RISC-V devices. Due to limited resources, cross compilation is used now for building OpenVINO targeting RISC-V development boards.

Reviewer 2: ```suggestion
This guide shows how to build OpenVINO Runtime for 64-bit RISC-V devices using cross-compilation.
```

I'm not sure - should we leave some words about reason of cross-compilation since it's just build tutorial 🤔

Author: If we will write another separate doc to introduce the build on RISC-V host (maybe `build_linux.md` is enough 🤔), we do not need to show the reasons here. If all RISC-V related tutorial is contained in this file, maybe the reason is necessary to show the current progress. 🤔

Cross compilation was tested on the following hosts:
- Ubuntu 22.04 (64-bit), x64

The software was validated on the following devices:
- [Lichee Pi 4A](https://wiki.sipeed.com/hardware/en/lichee/th1520/lp4a.html) with RVV 0.7.1
- [Banana Pi BPI-F3](https://www.banana-pi.org/en/banana-pi-sbcs/175.html) with RVV 1.0

## Software requirements

- [CMake](https://cmake.org/download/) 3.13 or higher
- GCC 7.5 or higher (for non-RVV) / [xuantie-gnu-toolchain](https://github.com/XUANTIE-RV/xuantie-gnu-toolchain) (for RVV)

> - Python 3.10 for OpenVINO Runtime Python API

Reviewer: Python 3.8 or higher

Author: Currently only Python 3.10 is tested, and there is something wrong with Python 3.11. So I am not sure if it is suitable here to mention "Python 3.8 or higher" 🤔

## How to build
### Build with RVV
0. Prerequisite - build `xuantie-gnu-toolchain` and `qemu`:
   ```sh
   git clone https://github.com/XUANTIE-RV/xuantie-gnu-toolchain.git
   cd xuantie-gnu-toolchain
   ./configure --prefix=<xuantie_install_path>
   make linux build-qemu -j$(nproc)
   ```
   > **NOTE**: The `build-qemu` target is optional, as it is used to build the `qemu` simulator. However, it is recommended to build the `qemu` simulator, since it is much more convenient to validate the software on your host than on your devices.

1. Clone OpenVINO repository and init submodules:
   ```sh

>    git clone https://github.com/openvinotoolkit/openvino.git
>    cd openvino
>    git submodule update --init --recursive

Reviewer 2: ```suggestion
   git clone --recursive https://github.com/openvinotoolkit/openvino.git
```
The key ` --recursive`  is the same `git submodule update --init --recursive`

   ```

2. Install build dependencies using the `install_build_dependencies.sh` script in the
   project root folder.
   ```sh
   sudo ./install_build_dependencies.sh
   ```

3. Create a build folder:
   ```sh
   mkdir build && cd build
   ``` 

4. To cross compile OpenVINO Runtime for RISC-V devices, run `cmake` with specified `CMAKE_TOOLCHAIN_FILE` and `RISCV_TOOLCHAIN_ROOT`:
   ```sh
   cmake .. \

>      -DCMAKE_BUILD_TYPE=Release \

Reviewer 2: The same: this is optional key. Could you please remove it?

Author: So you mean the key `-DCMAKE_BUILD_TYPE` is also optional?

Reviewer 2: In general, it is. Moreover, `Release` type is chosen by default.
But I don't insist

     -DCMAKE_INSTALL_PREFIX=<openvino_install_path> \
     -DENABLE_INTEL_CPU=ON \
     -DENABLE_INTEL_GPU=OFF \
     -DENABLE_INTEL_NPU=OFF \
     -DENABLE_AUTO=ON \
     -DENABLE_HETERO=ON \
     -DENABLE_MULTI=ON \

>      -DENABLE_JS=OFF \

Reviewer 2: I think we can skip these keys since `AUTO`, `HETERO` and `MULTI` are built by default. As for `-DENABLE_JS=OFF`, I suggested you to disable this target just to reduce build time.
In general it should work and not mentioned in guide.
Could you please remove them?

Author: Sure!

Reviewer: GPU should be disabled on cmake level. Our cmake scripts should have reasonable defaults to avoid such long list of non-default configuration values.

I ideally, I need no options except toolchains to build default OpenVINO for RISC-V

Author: So should I set these keys like `-DENABLE_INTEL_CPU=ON -DENABLE_INTEL_GPU=OFF -DENABLE_INTEL_NPU=OFF` inside toolchain files intead of in the command line?

Reviewer: we have some default values in cake/features.cmake file. You can adjust it there

Author: Thank you for telling me that! I will remove these non-default keys here then!

     -DCMAKE_TOOLCHAIN_FILE=../cmake/toolchains/<toolchain_file> \
     -DRISCV_TOOLCHAIN_ROOT=<xuantie_install_path> \
     -DENABLE_TESTS=ON \

>      -DENABLE_SAMPLES=ON

Reviewer 2: These keys are optional too. Could you remove them please?
It would be great if the guide describes only needed cmake keys

   ```
   > **NOTE**: To build OpenVINO Runtime for different versions of RVV, you just need to specify corresponding toolchain files. For exmaple, you can replace `<toolchain_file>` with `riscv64-071-thead-gnu.toolchain.cmake` for RVV 0.7.1 and `riscv64-100-thead-gnu.toolchain.cmake` for RVV 1.0 respectively.

   Then run `make` to build the project:
   ```sh
   make install -j$(nproc)
   ```

### Build without RVV
0. Prerequisite - build `riscv-gnu-toolchain`:
   ```sh
   git clone https://github.com/riscv-collab/riscv-gnu-toolchain.git
   cd riscv-gnu-toolchain

>    ./configure --prefix=/opt/riscv

Reviewer 2: ```suggestion
   ./configure --prefix=<xuantie_install_path>
```

Author: The same reason as above,  usually the `riscv-gnu-toolchain` is installed under `/opt/riscv` (refer [here](https://github.com/riscv-collab/riscv-gnu-toolchain?tab=readme-ov-file#installation-linux)), so I deliberately leave this path here.

   make linux build-qemu -j$(nproc)
   ```
   > **NOTE**: The `build-qemu` target is optional here as well. More information can be seen [here](https://github.com/riscv-collab/riscv-gnu-toolchain).

1. Clone OpenVINO repository and init submodules:
   ```sh
   git clone https://github.com/openvinotoolkit/openvino.git
   cd openvino
   git submodule update --init --recursive

>    ```

Reviewer: looks like we have copy paste with the previous toolchain approach. 
Maybe we can separate build steps on 2 stages: build toolchain(s) and then build OV with specific toolchain. Last step will be mentioned only once in this document.

Author: What you suggested does have a point! I will reorganize this part!

2. Install build dependencies using the `install_build_dependencies.sh` script in the
   project root folder.
   ```sh
   sudo ./install_build_dependencies.sh
   ```

3. Create a build folder:
   ```sh
   mkdir build && cd build
   ``` 

4. To cross compile OpenVINO Runtime for RISC-V devices, run `cmake` with specified `CMAKE_TOOLCHAIN_FILE` and `RISCV_TOOLCHAIN_ROOT`:
   ```sh
   cmake .. \
     -DCMAKE_BUILD_TYPE=Release \
     -DCMAKE_INSTALL_PREFIX=<openvino_install_path> \
     -DENABLE_INTEL_CPU=ON \
     -DENABLE_INTEL_GPU=OFF \
     -DENABLE_INTEL_NPU=OFF \
     -DENABLE_AUTO=ON \
     -DENABLE_HETERO=ON \
     -DENABLE_MULTI=ON \
     -DENABLE_JS=OFF \

>      -DCMAKE_TOOLCHAIN_FILE=../cmake/toolchains/riscv64-071-thead-gnu.toolchain.cmake \

Reviewer 2: This cmake toolchain file is used only for RVV0.7.1. Why is it mentioned in the section `Build without RVV`?
Can we reuse the existing default [toolchain](https://github.com/openvinotoolkit/openvino/blob/239466ca5d5333af11ee750260215941737383b3/cmake/toolchains/riscv64-gnu.toolchain.cmake)? What do you think?

Author: Yes, we should use the general toolchain file here. Thank you!

>      -DRISCV_TOOLCHAIN_ROOT=/opt/riscv \

Reviewer 2: ```suggestion
     -DRISCV_TOOLCHAIN_ROOT=<xuantie_install_path> \
```

Author: I am introducing the way of building OV targeting RISC-V devices without RVV here. Usually the `riscv-gnu-toolchain` is installed under `/opt/riscv` (refer [here](https://github.com/riscv-collab/riscv-gnu-toolchain?tab=readme-ov-file#installation-linux)), so I deliberately leave this path here.

     -DENABLE_TESTS=ON \
     -DENABLE_SAMPLES=ON
   ```
   > **NOTE**: The `riscv-gnu-toolchain` is build as there are essential files used for cross compilation under `/opt/riscv/sysroot`. The latest stable versions of Clang or GCC both support compiling source code into RISC-V instructions, so it is acceptable to choose your preferable compilers by specifying `-DCMAKE_C_COMPILER` and `CMAKE_CXX_COMPILER`. But remember to add the key `-DCMAKE_SYSROOT=/opt/riscv/sysroot`, otherwise many fundamental headers and libs could not be found during cross compilation. 

   Then run `make` to build the project:
   ```sh
   make install -j$(nproc)
   ```

> ### Build the OpenVINO Runtime Python API

Reviewer: can we merge this with main build steps and mark this as optional? 
E.g.:
1. Build toolchain(s)
2. (Optional) install python dependencies
3. Build OpenVINO

Author: Sure! Thanks for the suggestion!

1. Add riscv64 repositories to download riscv64-specific packages:
    ```sh
    echo deb [arch=riscv64] http://ports.ubuntu.com/ubuntu-ports/ jammy main >> riscv64-sources.list
    echo deb [arch=riscv64] http://ports.ubuntu.com/ubuntu-ports/ jammy universe >> riscv64-sources.list
    echo deb [arch=riscv64] http://ports.ubuntu.com/ubuntu-ports/ jammy-updates main >> riscv64-sources.list
    echo deb [arch=riscv64] http://ports.ubuntu.com/ubuntu-ports/ jammy-security main >> riscv64-sources.list
    mv riscv64-sources.list /etc/apt/sources.list.d/
    dpkg --add-architecture riscv64
    apt-get update -o Dir::Etc::sourcelist=/etc/apt/sources.list.d/riscv64-sources.list
    ```

2. Install `libpython3-dev:riscv64` using `apt-get`:
    ```sh
    apt-get install -y --no-install-recommends libpython3-dev:riscv64
    ```
   Create  symbolink to allow python to find `riscv64-linux-gnu/python3.10/pyconfig.h` in `/usr/include/python3.10/` (this header is initially stored in `/usr/include/riscv64-linux-gnu/`)
    ```sh
    ln -s /usr/include/riscv64-linux-gnu/ /usr/include/python3.10/

>     ```

Reviewer 2: What is about to move this part to `Software Requirements`? There is nothing about `libpython3-dev:riscv64` in that section 🤔 
What do you think?

Author: I think maybe leave this part here might be better. This part is actually optional and not *required*, so when you want to enable python API you could jump directly to this part and go through the whole process.

> 3. Add the keys `-DENABLE_PYTHON=ON -DENABLE_PYTHON_PACKAGING=ON -DENABLE_WHEEL=ON` to cmake command during OpenVINO build.

Reviewer: `-DENABLE_PYTHON_PACKAGING=ON` is required only if you build Debain package.
I think you can drop it here.

Author: Thanks for the advice!

> **Note**: Currently only Python 3.10 on Ubuntu 22.04 is verified. So the target device must have Python 3.10 in this case.

> ### How to use OpenVINO builds

Reviewer 2: I think this section is extra because this guide is not about that 😄 
Then we can leave only host related part:
```
### RISC-V Emulation software
In order to test applications without hardware one can use emulation software. The command line example to launch executable file with riscv64 emulation:
  <xuantie_install_path>/bin/qemu-riscv64 -cpu=<target_cpu> <executable_file_path>

  For example, to emulate RVV0.7.1:
  <xuantie_install_path>/bin/qemu-riscv64 -cpu rv64,v=true,vext_spec=v0.7.1 <executable_file_path>
  
Or to emulate RVV 1.0:
  <xuantie_install_path>/bin/qemu-riscv64 -cpu rv64,v=true,vext_spec=v1.0 <executable_file_path>
```
Will these keys `-cpu` work?

Author: It seems `rv64,x-v=true,vext_spec=v1.0` will work for xuantie-qemu.
